File: Leetcode 56 - merge-intervals/56.merge-intervals.cpp
/*
 * @lc app=leetcode id=56 lang=cpp
 *
 * [56] Merge Intervals
 */

// @lc code=start
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
  vector<vector<int>> merge(vector<vector<int>> &intervals) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    sort(intervals.begin(), intervals.end());

    if (intervals.empty())
      return {};

    vector<vector<int>> newNums;
    int start = intervals[0][0];
    int end = intervals[0][1];

    for (int i = 1; i < intervals.size(); i++) {
      if (intervals[i][0] <= end) {
        end = max(end, intervals[i][1]);
      } else {
        newNums.push_back({start, end});
        start = intervals[i][0];
        end = intervals[i][1];
      }
    }
    newNums.push_back({start, end});
    return newNums;
  }
};

int main() {
  Solution sol;
  vector<vector<int>> nums1 = {
      {1, 3}, {2, 6}, {8, 10}, {15, 18}}; // {{1, 6}, {8, 10}, {15, 18}}
  vector<vector<int>> result1 = sol.merge(nums1);

  for (vector<int> nums : result1) {
    cout << "{";
    for (const int &num : nums) {
      cout << num << ",";
    }
    cout << "}";
  }
  cout << endl;

  vector<vector<int>> nums2 = {{1, 4}, {4, 5}}; // {{1, 5}}
  vector<vector<int>> result2 = sol.merge(nums2);

  for (vector<int> nums : result2) {
    cout << "{";
    for (const int &num : nums) {
      cout << num << ",";
    }
    cout << "}";
  }
  cout << endl;
}
// @lc code=end

File: Leetcode 56 - merge-intervals/56.merge-intervals.java
/*
 * @lc app=leetcode id=56 lang=java
 *
 * [56] Merge Intervals
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {

  public int[][] merge(int[][] intervals) {
    Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
    List<int[]> newIntervals = new ArrayList<>();
    newIntervals.add(intervals[0]);

    for (int i = 1; i < intervals.length; i++) {
      int lastEnd = newIntervals.getLast()[1];

      // Combine if the end_i is larger than start_j
      // Update the end_i to the maximum of end_i and end_j
      if (lastEnd >= intervals[i][0]) {
        newIntervals.getLast()[1] = Math.max(lastEnd, intervals[i][1]);
      } else {
        newIntervals.add(intervals[i]);
      }
    }

    return newIntervals.toArray(new int[newIntervals.size()][]);
  }

  // Time Complexity: O(n log n)
  // Time Complexity: O(n)

  public static void main(String[] args) {
    Solution solution = new Solution();

    int[][] intervals = { { 1, 3 }, { 2, 6 }, { 8, 10 }, { 15, 18 } };
    int[][] result = solution.merge(intervals);
    System.out.println(Arrays.deepToString(result)); // [[1, 6], [8, 10], [15, 18]]

    intervals = new int[][] { { 1, 4 }, { 4, 5 } };
    result = solution.merge(intervals);
    System.out.println(Arrays.deepToString(result)); // [[1, 5]]

    intervals = new int[][] { { 4, 7 }, { 1, 4 } };
    result = solution.merge(intervals);
    System.out.println(Arrays.deepToString(result)); // [[1, 7]]

    intervals =
      new int[][] { { 2, 3 }, { 4, 5 }, { 6, 7 }, { 8, 9 }, { 1, 10 } };
    result = solution.merge(intervals);
    System.out.println(Arrays.deepToString(result)); // [[1, 10]]
  }
}
// @lc code=end
